# Session 19 GroupBy Object in Pandas

In [1]:
import numpy as np
import pandas as pd

In [2]:
movies = pd.read_csv("imdb-top-1000.csv")
movies.head(2)

Series_Title Released_Year  Runtime  Genre  IMDB_Rating  \
0  The Shawshank Redemption          1994      142  Drama          9.3   
1             The Godfather          1972      175  Crime          9.2   

               Director          Star1  No_of_Votes        Gross  Metascore  
0        Frank Darabont    Tim Robbins      2343110   28341469.0       80.0  
1  Francis Ford Coppola  Marlon Brando      1620367  134966411.0      100.0

## **GroupBy Object in Pandas**

### **Introduction**

The `groupby()` function in pandas is one of the most powerful tools for **data aggregation and analysis**. It allows us to **split the data into groups** based on certain column values, apply operations independently to each group, and then combine the results.

---

### **Concept**

* The **GroupBy** operation involves **splitting**, **applying**, and **combining**:

  1. **Split** – Divide the dataset into groups based on one or more **categorical columns**.
  2. **Apply** – Perform operations (like `sum()`, `mean()`, `count()`, etc.) on each group.
  3. **Combine** – Merge the results back into a new structure (often a DataFrame).

---

### **How Grouping Works**

When we call:

```python
df.groupby('column_name')
```

* Pandas creates a **`DataFrameGroupBy`** object.
* This object **does not return a subset of the DataFrame** immediately.
* Instead, it acts as an **intermediate object** that stores information about how the data is split internally.
* We can then perform **aggregate**, **filter**, or **transform** operations on this object.

---

### **Grouping Basis**

Grouping is generally done based on **categorical columns**, since:

* **Categorical columns** represent groups, categories, or labels (like "Gender", "Region", "Department").
* **Numerical columns** contain data on which we can perform aggregations (like sum, mean, max, etc.).

For example:

```python
import pandas as pd

data = {
    'Department': ['Sales', 'Sales', 'HR', 'HR', 'IT'],
    'Employee': ['A', 'B', 'C', 'D', 'E'],
    'Salary': [50000, 55000, 40000, 42000, 60000]
}

df = pd.DataFrame(data)

grouped = df.groupby('Department')
print(grouped)
```

**Output:**

```
<pandas.core.groupby.generic.DataFrameGroupBy object at 0x...>
```

This output confirms that `grouped` is a **GroupBy object**, not a DataFrame.

---

### **Accessing Aggregated Results**

We can perform operations on the grouped data:

```python
# Get mean salary of each department
df.groupby('Department')['Salary'].mean()
```

**Output:**

```
Department
HR        41000.0
IT        60000.0
Sales     52500.0
Name: Salary, dtype: float64
```

---

### **Summary**

| Aspect                | Description                                                                         |
| --------------------- | ----------------------------------------------------------------------------------- |
| **Purpose**           | To group data based on categorical values and perform operations on numerical data. |
| **Returned Object**   | `DataFrameGroupBy` (not a DataFrame itself)                                         |
| **Common Operations** | `sum()`, `mean()`, `count()`, `max()`, `min()`, `aggregate()`                       |
| **Grouping Basis**    | Usually on categorical columns.                                                     |

## Some Tasks using GroupBy

### **Task 1**: Find the top 3 genres by total earning, i.e. `Gross`

In [ ]:
# strategy 1
movies.groupby('Genre').sum().sort_values(by='Gross', ascending=False).head(3)['Gross']

Genre
Drama     3.540997e+10
Action    3.263226e+10
Comedy    1.566387e+10
Name: Gross, dtype: float64

In [32]:
# strategy 2 (faster)
movies.groupby('Genre')['Gross'].sum().sort_values(ascending=False).head(3)

Genre
Drama     3.540997e+10
Action    3.263226e+10
Comedy    1.566387e+10
Name: Gross, dtype: float64

The second strategy is faster because it filters out the required column first and then performs aggregations on the filtered or selected columns.

### **Task 2**: Find the genre with highest avg IMDB Rating
Following the efficient strategy in this one as discussed above by Selecting the **IMDB_Rating** column beforehand.

In [34]:
movies.groupby('Genre')['IMDB_Rating'].sum().sort_values(ascending=False).head(1)

Genre
Drama    2299.7
Name: IMDB_Rating, dtype: float64

### **Task 3**: Find the director with the most popularity.
Popularity described by **No_of_Votes** attribute.

In [42]:
movies.groupby('Director')['No_of_Votes'].sum().sort_values(ascending=False)

Director
Christopher Nolan    11578345
Quentin Tarantino     8123208
Steven Spielberg      7817166
David Fincher         6607859
Martin Scorsese       6513530
                       ...   
J. Lee Thompson         26457
Peter Mullan            25938
René Laloux             25229
Francis Lee             25198
Kaige Chen              25088
Name: No_of_Votes, Length: 548, dtype: int64

### **Task 4**: Find the highest rated movies of each genre.

In [58]:
movies.groupby('Genre')[['Genre', 'Series_Title', 'IMDB_Rating']].head(1)

Genre                     Series_Title  IMDB_Rating
0        Drama         The Shawshank Redemption          9.3
1        Crime                    The Godfather          9.2
2       Action                  The Dark Knight          9.0
7    Biography                 Schindler's List          8.9
12     Western  Il buono, il brutto, il cattivo          8.8
19      Comedy                     Gisaengchung          8.6
21   Adventure                     Interstellar          8.6
23   Animation    Sen to Chihiro no kamikakushi          8.6
49      Horror                           Psycho          8.5
69     Mystery                          Memento          8.4
309  Film-Noir                    The Third Man          8.1
321    Fantasy     Das Cabinet des Dr. Caligari          8.1
688     Family       E.T. the Extra-Terrestrial          7.8
700   Thriller                  Wait Until Dark          7.8

In [75]:
# or using this strategy, which is also the better one
movies.groupby('Genre')[['Series_Title', 'IMDB_Rating']].first()

Series_Title  IMDB_Rating
Genre                                                  
Action                     The Dark Knight          9.0
Adventure                     Interstellar          8.6
Animation    Sen to Chihiro no kamikakushi          8.6
Biography                 Schindler's List          8.9
Comedy                        Gisaengchung          8.6
Crime                        The Godfather          9.2
Drama             The Shawshank Redemption          9.3
Family          E.T. the Extra-Terrestrial          7.8
Fantasy       Das Cabinet des Dr. Caligari          8.1
Film-Noir                    The Third Man          8.1
Horror                              Psycho          8.5
Mystery                            Memento          8.4
Thriller                   Wait Until Dark          7.8
Western    Il buono, il brutto, il cattivo          8.8

### **Task 5**: Find the number of movies done by each actor.

In [ ]:
# simple strategy without groupby
movies['Star1'].value_counts()

Star1
Tom Hanks              12
Robert De Niro         11
Al Pacino              10
Clint Eastwood         10
Humphrey Bogart         9
                       ..
Richard Farnsworth      1
Junko Iwao              1
Fernanda Montenegro     1
Eli Marienthal          1
Preity Zinta            1
Name: count, Length: 660, dtype: int64

In [65]:
movies.groupby('Star1')['Series_Title'].count().sort_values(ascending=False)

Star1
Tom Hanks             12
Robert De Niro        11
Al Pacino             10
Clint Eastwood        10
Humphrey Bogart        9
                      ..
Günes Sensoy           1
Haluk Bilginer         1
Harriet Andersson      1
Harry Dean Stanton     1
Griffin Dunne          1
Name: Series_Title, Length: 660, dtype: int64

#### Assigning variable for the group for simplification of code ahead.

In [4]:
genres = movies.groupby('Genre')

In [6]:
# applying builtin aggregation functions on groupby objects
genres.sum()

Series_Title  \
Genre                                                          
Action     The Dark KnightThe Lord of the Rings: The Retu...   
Adventure  InterstellarBack to the FutureInglourious Bast...   
Animation  Sen to Chihiro no kamikakushiThe Lion KingHota...   
Biography  Schindler's ListGoodfellasHamiltonThe Intoucha...   
Comedy     GisaengchungLa vita è bellaModern TimesCity Li...   
Crime      The GodfatherThe Godfather: Part II12 Angry Me...   
Drama      The Shawshank RedemptionFight ClubForrest Gump...   
Family     E.T. the Extra-TerrestrialWilly Wonka & the Ch...   
Fantasy                Das Cabinet des Dr. CaligariNosferatu   
Film-Noir   The Third ManThe Maltese FalconShadow of a Doubt   
Horror     PsychoAlienThe ThingThe ExorcistNight of the L...   
Mystery    MementoRear WindowVertigoShutter IslandKahaani...   
Thriller                                     Wait Until Dark   
Western    Il buono, il brutto, il cattivoOnce Upon a Tim...   

                                               Released_Year  Runtime  \
Genre                                                                   
Action     2008200320102001200219991980197719621954200019...    22196   
Adventure  2014198520091981196819621959201319751963194819...     9656   
Animation  2001199419882016201820172008199719952019200920...     8166   
Biography  1993199020202011200220171995198420182013201320...    11970   
Comedy     2019199719361931200919641940200120001973196019...    17380   
Crime      1972197419571994200219991995199120192006199519...    13524   
Drama      1994199919941975202019981946201420061998198819...    36049   
Family                                              19821971      215   
Fantasy                                             19201922      170   
Film-Noir                                       194919411943      312   
Horror          19601979198219731968196120171978193320042001     1123   
Mystery     200019541958201020121995197219381988201219981997     1429   
Thriller                                                1967      108   
Western                                     1966196819651976      593   

           IMDB_Rating                                           Director  \
Genre                                                                       
Action          1367.3  Christopher NolanPeter JacksonChristopher Nola...   
Adventure        571.5  Christopher NolanRobert ZemeckisQuentin Tarant...   
Animation        650.3  Hayao MiyazakiRoger AllersIsao TakahataMakoto ...   
Biography        698.6  Steven SpielbergMartin ScorseseThomas KailOliv...   
Comedy          1224.7  Bong Joon HoRoberto BenigniCharles ChaplinChar...   
Crime            857.8  Francis Ford CoppolaFrancis Ford CoppolaSidney...   
Drama           2299.7  Frank DarabontDavid FincherRobert ZemeckisMilo...   
Family            15.6                         Steven SpielbergMel Stuart   
Fantasy           16.0                            Robert WieneF.W. Murnau   
Film-Noir         23.9              Carol ReedJohn HustonAlfred Hitchcock   
Horror            87.0  Alfred HitchcockRidley ScottJohn CarpenterWill...   
Mystery           95.7  Christopher NolanAlfred HitchcockAlfred Hitchc...   
Thriller           7.8                                      Terence Young   
Western           33.4  Sergio LeoneSergio LeoneSergio LeoneClint East...   

                                                       Star1  No_of_Votes  \
Genre                                                                       
Action     Christian BaleElijah WoodLeonardo DiCaprioElij...     72282412   
Adventure  Matthew McConaugheyMichael J. FoxBrad PittJürg...     22576163   
Animation  Daveigh ChaseRob MinkoffTsutomu TatsumiRyûnosu...     21978630   
Biography  Liam NeesonRobert De NiroLin-Manuel MirandaÉri...     24006844   
Comedy     Kang-ho SongRoberto BenigniCharles ChaplinChar...     27620327   
Crime      Marlon BrandoAl PacinoHenry FondaJohn Travolta...     33533615   
Drama

In [7]:
genres.min()

Series_Title Released_Year  Runtime  \
Genre                                                               
Action                                 300          1924       45   
Adventure            2001: A Space Odyssey          1925       88   
Animation                            Akira          1940       71   
Biography                 12 Years a Slave          1928       93   
Comedy                (500) Days of Summer          1921       68   
Crime                         12 Angry Men          1931       80   
Drama                                 1917          1925       64   
Family          E.T. the Extra-Terrestrial          1971      100   
Fantasy       Das Cabinet des Dr. Caligari          1920       76   
Film-Noir                Shadow of a Doubt          1941      100   
Horror                               Alien          1933       71   
Mystery                          Dark City          1938       96   
Thriller                   Wait Until Dark          1967      108   
Western    Il buono, il brutto, il cattivo          1965      132   

           IMDB_Rating               Director                     Star1  \
Genre                                                                     
Action             7.6       Abhishek Chaubey                Aamir Khan   
Adventure          7.6         Akira Kurosawa                Aamir Khan   
Animation          7.6            Adam Elliot             Adrian Molina   
Biography          7.6             Adam McKay              Adrien Brody   
Comedy             7.6  Alejandro G. Iñárritu                Aamir Khan   
Crime              7.6         Akira Kurosawa                Ajay Devgn   
Drama              7.6             Aamir Khan                Abhay Deol   
Family             7.8             Mel Stuart               Gene Wilder   
Fantasy            7.9            F.W. Murnau               Max Schreck   
Film-Noir          7.8       Alfred Hitchcock           Humphrey Bogart   
Horror             7.6     Alejandro Amenábar           Anthony Perkins   
Mystery            7.6            Alex Proyas  Bernard-Pierre Donnadieu   
Thriller           7.8          Terence Young            Audrey Hepburn   
Western            7.8         Clint Eastwood            Clint Eastwood   

           No_of_Votes        Gross  Metascore  
Genre                                           
Action           25312       3296.0       33.0  
Adventure        29999      61001.0       41.0  
Animation        25229     128985.0       61.0  
Biography        27254      21877.0       48.0  
Comedy           26337       1305.0       45.0  
Crime            27712       6013.0       47.0  
Drama            25088       3600.0       28.0  
Family          178731    4000000.0       67.0  
Fantasy          57428  337574718.0        NaN  
Film-Noir        59556     449191.0       94.0  
Horror           27007      89029.0       46.0  
Mystery          33982    1035953.0       52.0  
Thriller         27733   17550741.0       81.0  
Western          65659    5321508.0       69.0

### What does the above aggregation method do?
The above aggregation method when applied on the **GroupBy Object**, it sums up the value of each genre with the corresponding columns. *eg*: the sum of **IMDB_Rating** for all the movies of the action genre is **1367.3**.

Now here is another example, where we're using the `min()` aggregate function to identify the earliest release of a movie in specific genres, as well as also giving the min of other columns.

### **Task 6**: Count the number of groups formed by using `groupby()` on **Genre** column.

In [66]:
len(movies.groupby('Genre'))

14

In [68]:
# or this way
movies['Genre'].nunique()

14

### **Task 7**: Find the number of rows within each group.

In [69]:
movies.groupby('Genre').size()

Genre
Action       172
Adventure     72
Animation     82
Biography     88
Comedy       155
Crime        107
Drama        289
Family         2
Fantasy        2
Film-Noir      3
Horror        11
Mystery       12
Thriller       1
Western        4
dtype: int64

In [70]:
# or this way
movies.Genre.value_counts()

Genre
Drama        289
Action       172
Comedy       155
Crime        107
Biography     88
Animation     82
Adventure     72
Mystery       12
Horror        11
Western        4
Film-Noir      3
Fantasy        2
Family         2
Thriller       1
Name: count, dtype: int64

In [ ]:
# storing the groupby object in var for further code ahead
genres = movies.groupby('Genre')

### **Task 8**: Find the first, last, and nth item of the Group.

In [79]:
genres.first()
genres.last()
genres.nth(6)

Series_Title Released_Year  Runtime  \
16   Star Wars: Episode V - The Empire Strikes Back          1980      124   
27                                            Se7en          1995      127   
32                            It's a Wonderful Life          1946      130   
66                                           WALL·E          2008       98   
83                               The Great Dictator          1940      125   
102                                      Braveheart          1995      178   
118                              North by Northwest          1959      136   
420                                          Sleuth          1972      138   
724                                         Get Out          2017      104   

         Genre  IMDB_Rating              Director             Star1  \
16      Action          8.7        Irvin Kershner       Mark Hamill   
27       Crime          8.6         David Fincher    Morgan Freeman   
32       Drama          8.6           Frank Capra     James Stewart   
66   Animation          8.4        Andrew Stanton         Ben Burtt   
83      Comedy          8.4       Charles Chaplin   Charles Chaplin   
102  Biography          8.3            Mel Gibson        Mel Gibson   
118  Adventure          8.3      Alfred Hitchcock        Cary Grant   
420    Mystery          8.0  Joseph L. Mankiewicz  Laurence Olivier   
724     Horror          7.7          Jordan Peele    Daniel Kaluuya   

     No_of_Votes        Gross  Metascore  
16       1159315  290475067.0       82.0  
27       1445096  100125643.0       65.0  
32        405801   82385199.0       89.0  
66        999790  223808164.0       95.0  
83        203150     288475.0        NaN  
102       959181   75600000.0       68.0  
118       299198   13275000.0       98.0  
420        44748    4081254.0        NaN  
724       492851  176040665.0       85.0

## `get_group()`
Returns a dataframe containing all the elements within a Group.

In [80]:
genres.get_group('Horror')

Series_Title Released_Year  Runtime   Genre  IMDB_Rating  \
49                     Psycho          1960      109  Horror          8.5   
75                      Alien          1979      117  Horror          8.4   
271                 The Thing          1982      109  Horror          8.1   
419              The Exorcist          1973      122  Horror          8.0   
544  Night of the Living Dead          1968       96  Horror          7.9   
707             The Innocents          1961      100  Horror          7.8   
724                   Get Out          2017      104  Horror          7.7   
844                 Halloween          1978       91  Horror          7.7   
876         The Invisible Man          1933       71  Horror          7.7   
932                       Saw          2004      103  Horror          7.6   
948                The Others          2001      101  Horror          7.6   

               Director             Star1  No_of_Votes        Gross  Metascore  
49     Alfred Hitchcock   Anthony Perkins       604211   32000000.0       97.0  
75         Ridley Scott  Sigourney Weaver       787806   78900000.0       89.0  
271      John Carpenter      Kurt Russell       371271   13782838.0       57.0  
419    William Friedkin     Ellen Burstyn       362393  232906145.0       81.0  
544    George A. Romero       Duane Jones       116557      89029.0       89.0  
707        Jack Clayton      Deborah Kerr        27007    2616000.0       88.0  
724        Jordan Peele    Daniel Kaluuya       492851  176040665.0       85.0  
844      John Carpenter  Donald Pleasence       233106   47000000.0       87.0  
876         James Whale      Claude Rains        30683  298791505.0       87.0  
932           James Wan        Cary Elwes       379020   56000369.0       46.0  
948  Alejandro Amenábar     Nicole Kidman       337651   96522687.0       74.0

Although we can also do the above mentioned thing by using the "filtering" strategy of native pandas in the following way:

In [81]:
movies[movies['Genre'] == 'Horror']

Series_Title Released_Year  Runtime   Genre  IMDB_Rating  \
49                     Psycho          1960      109  Horror          8.5   
75                      Alien          1979      117  Horror          8.4   
271                 The Thing          1982      109  Horror          8.1   
419              The Exorcist          1973      122  Horror          8.0   
544  Night of the Living Dead          1968       96  Horror          7.9   
707             The Innocents          1961      100  Horror          7.8   
724                   Get Out          2017      104  Horror          7.7   
844                 Halloween          1978       91  Horror          7.7   
876         The Invisible Man          1933       71  Horror          7.7   
932                       Saw          2004      103  Horror          7.6   
948                The Others          2001      101  Horror          7.6   

               Director             Star1  No_of_Votes        Gross  Metascore  
49     Alfred Hitchcock   Anthony Perkins       604211   32000000.0       97.0  
75         Ridley Scott  Sigourney Weaver       787806   78900000.0       89.0  
271      John Carpenter      Kurt Russell       371271   13782838.0       57.0  
419    William Friedkin     Ellen Burstyn       362393  232906145.0       81.0  
544    George A. Romero       Duane Jones       116557      89029.0       89.0  
707        Jack Clayton      Deborah Kerr        27007    2616000.0       88.0  
724        Jordan Peele    Daniel Kaluuya       492851  176040665.0       85.0  
844      John Carpenter  Donald Pleasence       233106   47000000.0       87.0  
876         James Whale      Claude Rains        30683  298791505.0       87.0  
932           James Wan        Cary Elwes       379020   56000369.0       46.0  
948  Alejandro Amenábar     Nicole Kidman       337651   96522687.0       74.0

## `groups` attribute of a **GroupBy Object**
This attribute or method returns a dictionary containing the **Groups** as **keys** and a list of indices as their corresponding **values**.

## `describe`, `sample`, `nunique`
These are some functions that we had discussed in Series as well as in DataFrames. These functions are also applicable on GroupBy Objects and the DataFrame formed out of them.

In [84]:
genres.describe()

Runtime                                                              \
            count        mean        std    min     25%    50%     75%    max   
Genre                                                                           
Action      172.0  129.046512  28.500706   45.0  110.75  127.5  143.25  321.0   
Adventure    72.0  134.111111  33.317320   88.0  109.00  127.0  149.00  228.0   
Animation    82.0   99.585366  14.530471   71.0   90.00   99.5  106.75  137.0   
Biography    88.0  136.022727  25.514466   93.0  120.00  129.0  146.25  209.0   
Comedy      155.0  112.129032  22.946213   68.0   96.00  106.0  124.50  188.0   
Crime       107.0  126.392523  27.689231   80.0  106.50  122.0  141.50  229.0   
Drama       289.0  124.737024  27.740490   64.0  105.00  121.0  137.00  242.0   
Family        2.0  107.500000  10.606602  100.0  103.75  107.5  111.25  115.0   
Fantasy       2.0   85.000000  12.727922   76.0   80.50   85.0   89.50   94.0   
Film-Noir     3.0  104.000000   4.000000  100.0  102.00  104.0  106.00  108.0   
Horror       11.0  102.090909  13.604812   71.0   98.00  103.0  109.00  122.0   
Mystery      12.0  119.083333  14.475423   96.0  110.75  117.5  130.25  138.0   
Thriller      1.0  108.000000        NaN  108.0  108.00  108.0  108.00  108.0   
Western       4.0  148.250000  17.153717  132.0  134.25  148.0  162.00  165.0   

          IMDB_Rating            ...         Gross              Metascore  \
                count      mean  ...           75%          max     count   
Genre                            ...                                        
Action          172.0  7.949419  ...  2.674437e+08  936662225.0     143.0   
Adventure        72.0  7.937500  ...  1.998070e+08  874211619.0      64.0   
Animation        82.0  7.930488  ...  2.520612e+08  873839108.0      75.0   
Biography        88.0  7.938636  ...  9.829924e+07  753585104.0      79.0   
Comedy          155.0  7.901290  ...  8.107809e+07  886752933.0     125.0   
Crime           107.0  8.016822  ...  7.102163e+07  790482117.0      87.0   
Drama           289.0  7.957439  ...  1.164461e+08  924558264.0     241.0   
Family            2.0  7.800000  ...  3.273329e+08  435110554.0       2.0   
Fantasy           2.0  8.000000  ...  4.182577e+08  445151978.0       0.0   
Film-Noir         3.0  7.966667  ...  6.273068e+07  123353292.0       3.0   
Horror           11.0  7.909091  ...  1.362817e+08  298791505.0      11.0   
Mystery          12.0  7.975000  ...  1.310949e+08  474203697.0       8.0   
Thriller          1.0  7.800000  ...  1.755074e+07   17550741.0       1.0   
Western           4.0  8.350000  ...  1.920000e+07   31800000.0       4.0   

                                                                  
                mean        std   min    25%   50%    75%    max  
Genre                                                             
Action     73.419580  12.421252  33.0  65.00  74.0  82.00   98.0  
Adventure  78.437500  12.345393  41.0  69.75  80.5  87.25  100.0  
Animation  81.093333   8.813646  61.0  75.00  82.0  87.50   96.0  
Biography  76.240506  11.028187  48.0  70.50  76.0  84.50   97.0  
Comedy     78.720000  11.829160  45.0  72.00  79.0  88.00   99.0  
Crime      77.080460  13.099102  47.0  69.50  77.0  87.00  100.0  
Drama      79.701245  12.744687  28.0  72.00  82.0  89.00  100.0  
Family     79.000000  16.970563  67.0  73.00  79.0  85.00   91.0  
Fantasy          NaN        NaN   NaN    NaN   NaN    NaN    NaN  
Film-Noir  95.666667   1.527525  94.0  95.00  96.0  96.50   97.0  
Horror     80.000000  15.362291  46.0  77.50  87.0  88.50   97.0  
Mystery    79.125000  18.604435  52.0  65.25  77.0  98.50  100.0  
Thriller   81.000000        NaN  81.0  81.00  81.0  81.00   81.0  
Western    78.250000   9.032349  69.0  72.75  77.0  82.50   90.0  

[14 rows x 40 columns]

In [ ]:
# 1 random values from each group
genres.sample()

Series_Title Released_Year  Runtime      Genre  \
152                  V for Vendetta          2005      132     Action   
675      Back to the Future Part II          1989      108  Adventure   
395  The Nightmare Before Christmas          1993       76  Animation   
650                     October Sky          1999      108  Biography   
780             Lost in Translation          2003      102     Comedy   
556            Strangers on a Train          1951      101      Crime   
297                   Jungfrukällan          1960       89      Drama   
688      E.T. the Extra-Terrestrial          1982      115     Family   
321    Das Cabinet des Dr. Caligari          1920       76    Fantasy   
712               Shadow of a Doubt          1943      108  Film-Noir   
419                    The Exorcist          1973      122     Horror   
961                    Lost Highway          1997      134    Mystery   
700                 Wait Until Dark          1967      108   Thriller   
691          The Outlaw Josey Wales          1976      135    Western   

     IMDB_Rating          Director            Star1  No_of_Votes        Gross  \
152          8.2    James McTeigue     Hugo Weaving      1032749   70511035.0   
675          7.8   Robert Zemeckis   Michael J. Fox       481918  118500000.0   
395          8.0      Henry Selick     Danny Elfman       300208   75082668.0   
650          7.8      Joe Johnston  Jake Gyllenhaal        82855   32481825.0   
780          7.7     Sofia Coppola      Bill Murray       415074   44585453.0   
556          7.9  Alfred Hitchcock   Farley Granger       123341    7630000.0   
297          8.1    Ingmar Bergman    Max von Sydow        26697    1526000.0   
688          7.8  Steven Spielberg     Henry Thomas       372490  435110554.0   
321          8.1      Robert Wiene    Werner Krauss        57428  337574718.0   
712          7.8  Alfred Hitchcock    Teresa Wright        59556  123353292.0   
419          8.0  William Friedkin    Ellen Burstyn       362393  232906145.0   
961          7.6       David Lynch     Bill Pullman       131101    3796699.0   
700          7.8     Terence Young   Audrey Hepburn        27733   17550741.0   
691          7.8    Clint Eastwood   Clint Eastwood        65659   31800000.0   

     Metascore  
152       62.0  
675       57.0  
395       82.0  
650       71.0  
780       89.0  
556       88.0  
297        NaN  
688       91.0  
321        NaN  
712       94.0  
419       81.0  
961       52.0  
700       81.0  
691       69.0

In [86]:
genres.nunique()

Series_Title  Released_Year  Runtime  IMDB_Rating  Director  Star1  \
Genre                                                                           
Action              172             61       78           15       123    121   
Adventure            72             49       58           10        59     59   
Animation            82             35       41           11        51     77   
Biography            88             44       56           13        76     72   
Comedy              155             72       70           11       113    133   
Crime               106             56       65           14        86     85   
Drama               289             83       95           14       211    250   
Family                2              2        2            1         2      2   
Fantasy               2              2        2            2         2      2   
Film-Noir             3              3        3            3         3      3   
Horror               11             11       10            8        10     11   
Mystery              12             11       10            8        10     11   
Thriller              1              1        1            1         1      1   
Western               4              4        4            4         2      2   

           No_of_Votes  Gross  Metascore  
Genre                                     
Action             172    172         50  
Adventure           72     72         33  
Animation           82     82         29  
Biography           88     88         40  
Comedy             155    155         44  
Crime              107    107         39  
Drama              288    287         52  
Family               2      2          2  
Fantasy              2      2          0  
Film-Noir            3      3          3  
Horror              11     11          9  
Mystery             12     12          7  
Thriller             1      1          1  
Western              4      4          4

In [92]:
genres['Released_Year'].max()

Genre
Action       2019
Adventure      PG
Animation    2020
Biography    2020
Comedy       2020
Crime        2019
Drama        2020
Family       1982
Fantasy      1922
Film-Noir    1949
Horror       2017
Mystery      2012
Thriller     1967
Western      1976
Name: Released_Year, dtype: object

## `agg()`
This method is used in cases where we'd want to apply different aggregate methods on different columns of the Group. Here, we pass a dictionary as a parameter, consisting of the column names as **Keys** and the function to be applied on them as their corresponding **Values**.

In [95]:
genres.agg(
    {
        'Runtime': 'mean',
        'IMDB_Rating': 'mean',
        'No_of_Votes': 'sum',
        'Gross': 'sum',
        'Metascore': 'min'
    }
)

Runtime  IMDB_Rating  No_of_Votes         Gross  Metascore
Genre                                                                   
Action     129.046512     7.949419     72282412  3.263226e+10       33.0
Adventure  134.111111     7.937500     22576163  9.496922e+09       41.0
Animation   99.585366     7.930488     21978630  1.463147e+10       61.0
Biography  136.022727     7.938636     24006844  8.276358e+09       48.0
Comedy     112.129032     7.901290     27620327  1.566387e+10       45.0
Crime      126.392523     8.016822     33533615  8.452632e+09       47.0
Drama      124.737024     7.957439     61367304  3.540997e+10       28.0
Family     107.500000     7.800000       551221  4.391106e+08       67.0
Fantasy     85.000000     8.000000       146222  7.827267e+08        NaN
Film-Noir  104.000000     7.966667       367215  1.259105e+08       94.0
Horror     102.090909     7.909091      3742556  1.034649e+09       46.0
Mystery    119.083333     7.975000      4203004  1.256417e+09       52.0
Thriller   108.000000     7.800000        27733  1.755074e+07       81.0
Western    148.250000     8.350000      1289665  5.822151e+07       69.0

On the other hand, using the same method, we can also generate a DF consisiting of the resultant of aggregate function for each distinct columns. This is as follows:

In [97]:
genres.agg(['min', 'max'])

Series_Title  \
                                       min   
Genre                                        
Action                                 300   
Adventure            2001: A Space Odyssey   
Animation                            Akira   
Biography                 12 Years a Slave   
Comedy                (500) Days of Summer   
Crime                         12 Angry Men   
Drama                                 1917   
Family          E.T. the Extra-Terrestrial   
Fantasy       Das Cabinet des Dr. Caligari   
Film-Noir                Shadow of a Doubt   
Horror                               Alien   
Mystery                          Dark City   
Thriller                   Wait Until Dark   
Western    Il buono, il brutto, il cattivo   

                                               Released_Year       Runtime  \
                                           max           min   max     min   
Genre                                                                        
Action                                 Yôjinbô          1924  2019      45   
Adventure                           Zombieland          1925    PG      88   
Animation          Ôkami kodomo no Ame to Yuki          1940  2020      71   
Biography                              Zerkalo          1928  2020      93   
Comedy                Zindagi Na Milegi Dobara          1921  2020      68   
Crime                        À bout de souffle          1931  2019      80   
Drama                                Zwartboek          1925  2020      64   
Family     Willy Wonka & the Chocolate Factory          1971  1982     100   
Fantasy                              Nosferatu          1920  1922      76   
Film-Noir                        The Third Man          1941  1949     100   
Horror                               The Thing          1933  2017      71   
Mystery                                Vertigo          1938  2012      96   
Thriller                       Wait Until Dark          1967  1967     108   
Western                 The Outlaw Josey Wales          1965  1976     132   

               IMDB_Rating                    Director                    \
           max         min  max                    min               max   
Genre                                                                      
Action     321         7.6  9.0       Abhishek Chaubey       Zack Snyder   
Adventure  228         7.6  8.6         Akira Kurosawa  Ömer Faruk Sorak   
Animation  137         7.6  8.6            Adam Elliot   Yoshifumi Kondô   
Biography  209         7.6  8.9             Adam McKay      Tom McCarthy   
Comedy     188         7.6  8.6  Alejandro G. Iñárritu       Zoya Akhtar   
Crime      229         7.6  9.2         Akira Kurosawa      Yavuz Turgul   
Drama      242         7.6  9.3             Aamir Khan       Çagan Irmak   
Family     115         7.8  7.8             Mel Stuart  Steven Spielberg   
Fantasy     94         7.9  8.1            F.W. Murnau      Robert Wiene   
Film-Noir  108         7.8  8.1       Alfred Hitchcock       John Huston   
Horror     122         7.6  8.5     Alejandro Amenábar  William Friedkin   
Mystery    138         7.6  8.4            Alex Proyas     Terry Gilliam   
Thriller   108         7.8  7.8          Terence Young     Terence Young   
Western    165         7.8  8.8         Clint Eastwood      Sergio Leone   

                              Star1                   No_of_Votes           \
                                min               max         min      max   
Genre                                                                        
Action                   Aamir Khan      Yun-Fat Chow       25312  2303232   
Adventure                Aamir Khan      Yves Montand       29999  1512360   
Animation             Adrian Molina      Yôji Matsuda       25229   999790   
Biography              Adrien Brody     Éric Toledano       27254  1213505   
Comedy                   Aamir Khan  Ömer Faruk Sorak       26337   939631   
Cri

Furthermore, we will merge both the syntaxes.

In [98]:
genres.agg(
    {
        'Runtime': ['min', 'mean'],
        'IMDB_Rating': 'mean',
        'No_of_Votes': 'sum',
        'Gross': 'sum',
        'Metascore': 'min'
    }
)

Runtime             IMDB_Rating No_of_Votes         Gross Metascore
              min        mean        mean         sum           sum       min
Genre                                                                        
Action         45  129.046512    7.949419    72282412  3.263226e+10      33.0
Adventure      88  134.111111    7.937500    22576163  9.496922e+09      41.0
Animation      71   99.585366    7.930488    21978630  1.463147e+10      61.0
Biography      93  136.022727    7.938636    24006844  8.276358e+09      48.0
Comedy         68  112.129032    7.901290    27620327  1.566387e+10      45.0
Crime          80  126.392523    8.016822    33533615  8.452632e+09      47.0
Drama          64  124.737024    7.957439    61367304  3.540997e+10      28.0
Family        100  107.500000    7.800000      551221  4.391106e+08      67.0
Fantasy        76   85.000000    8.000000      146222  7.827267e+08       NaN
Film-Noir     100  104.000000    7.966667      367215  1.259105e+08      94.0
Horror         71  102.090909    7.909091     3742556  1.034649e+09      46.0
Mystery        96  119.083333    7.975000     4203004  1.256417e+09      52.0
Thriller      108  108.000000    7.800000       27733  1.755074e+07      81.0
Western       132  148.250000    8.350000     1289665  5.822151e+07      69.0

## Looping on Groups

In [107]:
df = pd.DataFrame(columns=movies.columns)

for group, data in genres:
    df = pd.concat((df, data[data['IMDB_Rating'] == data['IMDB_Rating'].max()]))

df

/tmp/ipykernel_33788/1425520264.py:4: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  df = pd.concat((df, data[data['IMDB_Rating'] == data['IMDB_Rating'].max()]))


Series_Title Released_Year Runtime      Genre  \
2                        The Dark Knight          2008     152     Action   
21                          Interstellar          2014     169  Adventure   
23         Sen to Chihiro no kamikakushi          2001     125  Animation   
7                       Schindler's List          1993     195  Biography   
19                          Gisaengchung          2019     132     Comedy   
26                       La vita è bella          1997     116     Comedy   
1                          The Godfather          1972     175      Crime   
0               The Shawshank Redemption          1994     142      Drama   
688           E.T. the Extra-Terrestrial          1982     115     Family   
698  Willy Wonka & the Chocolate Factory          1971     100     Family   
321         Das Cabinet des Dr. Caligari          1920      76    Fantasy   
309                        The Third Man          1949     104  Film-Noir   
49                                Psycho          1960     109     Horror   
69                               Memento          2000     113    Mystery   
81                           Rear Window          1954     112    Mystery   
700                      Wait Until Dark          1967     108   Thriller   
12       Il buono, il brutto, il cattivo          1966     161    Western   

     IMDB_Rating              Director                Star1 No_of_Votes  \
2            9.0     Christopher Nolan       Christian Bale     2303232   
21           8.6     Christopher Nolan  Matthew McConaughey     1512360   
23           8.6        Hayao Miyazaki        Daveigh Chase      651376   
7            8.9      Steven Spielberg          Liam Neeson     1213505   
19           8.6          Bong Joon Ho         Kang-ho Song      552778   
26           8.6       Roberto Benigni      Roberto Benigni      623629   
1            9.2  Francis Ford Coppola        Marlon Brando     1620367   
0            9.3        Frank Darabont          Tim Robbins     2343110   
688          7.8      Steven Spielberg         Henry Thomas      372490   
698          7.8            Mel Stuart          Gene Wilder      178731   
321          8.1          Robert Wiene        Werner Krauss       57428   
309          8.1            Carol Reed         Orson Welles      158731   
49           8.5      Alfred Hitchcock      Anthony Perkins      604211   
69           8.4     Christopher Nolan           Guy Pearce     1125712   
81           8.4      Alfred Hitchcock        James Stewart      444074   
700          7.8         Terence Young       Audrey Hepburn       27733   
12           8.8          Sergio Leone       Clint Eastwood      688390   

           Gross  Metascore  
2    534858444.0       84.0  
21   188020017.0       74.0  
23    10055859.0       96.0  
7     96898818.0       94.0  
19    53367844.0       96.0  
26    57598247.0       59.0  
1    134966411.0      100.0  
0     28341469.0       80.0  
688  435110554.0       91.0  
698    4000000.0       67.0  
321  337574718.0        NaN  
309     449191.0       97.0  
49    32000000.0       97.0  
69    25544867.0       80.0  
81    36764313.0      100.0  
700   17550741.0       81.0  
12     6100000.0       90.0

## split -> apply -> combine
This simply refers to the viewpoint of the GroupBy method. This is diagramatically elaborated below.

          +--------------------+
          |     Original DF    |
          +--------------------+
                     │
                     ▼
             [Split by column]
                     │
         ┌───────────┼────────────┐
         ▼           ▼            ▼
      Group 1     Group 2      Group 3
         │           │            │
         ▼           ▼            ▼
     apply(sum)   apply(sum)   apply(sum)
         │           │            │
         └───────────┼────────────┘
                     ▼
             [Combine Results]
                     │
                     ▼
          +----------------------+
          |  Aggregated DataFrame |
          +----------------------+


In [142]:
genres.apply(min)

/tmp/ipykernel_33788/1443917609.py:1: FutureWarning: The provided callable <built-in function min> is currently using np.minimum.reduce. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string np.minimum.reduce instead.
  genres.apply(min)
/tmp/ipykernel_33788/1443917609.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  genres.apply(min)


Series_Title Released_Year  Runtime      Genre  \
Genre                                                                          
Action                                 300          1924       45     Action   
Adventure            2001: A Space Odyssey          1925       88  Adventure   
Animation                            Akira          1940       71  Animation   
Biography                 12 Years a Slave          1928       93  Biography   
Comedy                (500) Days of Summer          1921       68     Comedy   
Crime                         12 Angry Men          1931       80      Crime   
Drama                                 1917          1925       64      Drama   
Family          E.T. the Extra-Terrestrial          1971      100     Family   
Fantasy       Das Cabinet des Dr. Caligari          1920       76    Fantasy   
Film-Noir                Shadow of a Doubt          1941      100  Film-Noir   
Horror                               Alien          1933       71     Horror   
Mystery                          Dark City          1938       96    Mystery   
Thriller                   Wait Until Dark          1967      108   Thriller   
Western    Il buono, il brutto, il cattivo          1965      132    Western   

           IMDB_Rating               Director                     Star1  \
Genre                                                                     
Action             7.6       Abhishek Chaubey                Aamir Khan   
Adventure          7.6         Akira Kurosawa                Aamir Khan   
Animation          7.6            Adam Elliot             Adrian Molina   
Biography          7.6             Adam McKay              Adrien Brody   
Comedy             7.6  Alejandro G. Iñárritu                Aamir Khan   
Crime              7.6         Akira Kurosawa                Ajay Devgn   
Drama              7.6             Aamir Khan                Abhay Deol   
Family             7.8             Mel Stuart               Gene Wilder   
Fantasy            7.9            F.W. Murnau               Max Schreck   
Film-Noir          7.8       Alfred Hitchcock           Humphrey Bogart   
Horror             7.6     Alejandro Amenábar           Anthony Perkins   
Mystery            7.6            Alex Proyas  Bernard-Pierre Donnadieu   
Thriller           7.8          Terence Young            Audrey Hepburn   
Western            7.8         Clint Eastwood            Clint Eastwood   

           No_of_Votes        Gross  Metascore  
Genre                                           
Action           25312       3296.0        NaN  
Adventure        29999      61001.0        NaN  
Animation        25229     128985.0        NaN  
Biography        27254      21877.0        NaN  
Comedy           26337       1305.0        NaN  
Crime            27712       6013.0        NaN  
Drama            25088       3600.0        NaN  
Family          178731    4000000.0       67.0  
Fantasy          57428  337574718.0        NaN  
Film-Noir        59556     449191.0       94.0  
Horror           27007      89029.0       46.0  
Mystery          33982    1035953.0        NaN  
Thriller         27733   17550741.0       81.0  
Western          65659    5321508.0       69.0

### **Task 9**: Find the number of movies starting with the letter 'A' in each group.

In [143]:
def moviesWithA(group):
    return group['Series_Title'].str.startswith('A').sum()

In [144]:
genres.apply(moviesWithA)

/tmp/ipykernel_33788/6574545.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  genres.apply(moviesWithA)


Genre
Action       10
Adventure     2
Animation     2
Biography     9
Comedy       14
Crime         4
Drama        21
Family        0
Fantasy       0
Film-Noir     0
Horror        1
Mystery       0
Thriller      0
Western       0
dtype: int64

### **Task 10**: Find the ranking of each movie in the group of "Genres" according to IMDB Score.

In [145]:
def rank_movie(group):
    group['genre_rank'] = group['IMDB_Rating'].rank(ascending=False)
    return group

In [146]:
genres.apply(rank_movie)

/tmp/ipykernel_33788/2409710219.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  genres.apply(rank_movie)


Series_Title Released_Year  \
Genre                                                                           
Action   2                                      The Dark Knight          2008   
         5        The Lord of the Rings: The Return of the King          2003   
         8                                            Inception          2010   
         10   The Lord of the Rings: The Fellowship of the Ring          2001   
         13               The Lord of the Rings: The Two Towers          2002   
...                                                         ...           ...   
Thriller 700                                    Wait Until Dark          1967   
Western  12                     Il buono, il brutto, il cattivo          1966   
         48                        Once Upon a Time in the West          1968   
         115                         Per qualche dollaro in più          1965   
         691                             The Outlaw Josey Wales          1976   

              Runtime     Genre  IMDB_Rating           Director  \
Genre                                                             
Action   2        152    Action          9.0  Christopher Nolan   
         5        201    Action          8.9      Peter Jackson   
         8        148    Action          8.8  Christopher Nolan   
         10       178    Action          8.8      Peter Jackson   
         13       179    Action          8.7      Peter Jackson   
...               ...       ...          ...                ...   
Thriller 700      108  Thriller          7.8      Terence Young   
Western  12       161   Western          8.8       Sergio Leone   
         48       165   Western          8.5       Sergio Leone   
         115      132   Western          8.3       Sergio Leone   
         691      135   Western          7.8     Clint Eastwood   

                          Star1  No_of_Votes        Gross  Metascore  \
Genre                                                                  
Action   2       Christian Bale      2303232  534858444.0       84.0   
         5          Elijah Wood      1642758  377845905.0       94.0   
         8    Leonardo DiCaprio      2067042  292576195.0       74.0   
         10         Elijah Wood      1661481  315544750.0       92.0   
         13         Elijah Wood      1485555  342551365.0       87.0   
...                         ...          ...          ...        ...   
Thriller 700     Audrey Hepburn        27733   17550741.0       81.0   
Western  12      Clint Eastwood       688390    6100000.0       90.0   
         48         Henry Fonda       302844    5321508.0       80.0   
         115     Clint Eastwood       232772   15000000.0       74.0   
         691     Clint Eastwood        65659   31800000.0       69.0   

              genre_rank  
Genre                     
Action   2           1.0  
         5           2.0  
         8           3.5  
         10          3.5  
         13          6.0  
...                  ...  
Thriller 700         1.0  
Western  12          1.0  
         48          2.0  
         115         3.0  
         691         4.0  

[1000 rows x 11 columns]

In [147]:
genres['IMDB_Rating'].rank()

0      289.0
1      107.0
2      172.0
3      105.5
4      105.5
       ...  
995      9.0
996     17.5
997     17.5
998     17.5
999      7.0
Name: IMDB_Rating, Length: 1000, dtype: float64

### **Task 11**: Find normalized IMDB Rating group wise.

In [154]:
def normal(group):
    group['norm_rating'] = (group['IMDB_Rating'] - group['IMDB_Rating'].min()) / (group['IMDB_Rating'].max() - group['IMDB_Rating'].min())
    return group


In [155]:
genres.apply(normal)

/tmp/ipykernel_33788/1987004062.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  genres.apply(normal)


Series_Title Released_Year  \
Genre                                                                           
Action   2                                      The Dark Knight          2008   
         5        The Lord of the Rings: The Return of the King          2003   
         8                                            Inception          2010   
         10   The Lord of the Rings: The Fellowship of the Ring          2001   
         13               The Lord of the Rings: The Two Towers          2002   
...                                                         ...           ...   
Thriller 700                                    Wait Until Dark          1967   
Western  12                     Il buono, il brutto, il cattivo          1966   
         48                        Once Upon a Time in the West          1968   
         115                         Per qualche dollaro in più          1965   
         691                             The Outlaw Josey Wales          1976   

              Runtime     Genre  IMDB_Rating           Director  \
Genre                                                             
Action   2        152    Action          9.0  Christopher Nolan   
         5        201    Action          8.9      Peter Jackson   
         8        148    Action          8.8  Christopher Nolan   
         10       178    Action          8.8      Peter Jackson   
         13       179    Action          8.7      Peter Jackson   
...               ...       ...          ...                ...   
Thriller 700      108  Thriller          7.8      Terence Young   
Western  12       161   Western          8.8       Sergio Leone   
         48       165   Western          8.5       Sergio Leone   
         115      132   Western          8.3       Sergio Leone   
         691      135   Western          7.8     Clint Eastwood   

                          Star1  No_of_Votes        Gross  Metascore  \
Genre                                                                  
Action   2       Christian Bale      2303232  534858444.0       84.0   
         5          Elijah Wood      1642758  377845905.0       94.0   
         8    Leonardo DiCaprio      2067042  292576195.0       74.0   
         10         Elijah Wood      1661481  315544750.0       92.0   
         13         Elijah Wood      1485555  342551365.0       87.0   
...                         ...          ...          ...        ...   
Thriller 700     Audrey Hepburn        27733   17550741.0       81.0   
Western  12      Clint Eastwood       688390    6100000.0       90.0   
         48         Henry Fonda       302844    5321508.0       80.0   
         115     Clint Eastwood       232772   15000000.0       74.0   
         691     Clint Eastwood        65659   31800000.0       69.0   

              norm_rating  
Genre                      
Action   2       1.000000  
         5       0.928571  
         8       0.857143  
         10      0.857143  
         13      0.785714  
...                   ...  
Thriller 700          NaN  
Western  12      1.000000  
         48      0.700000  
         115     0.500000  
         691     0.000000  

[1000 rows x 11 columns]

## GroupBy on multiple columns
`Case`: Suppose we take a case where we want to bind the values from two columns and group them together such that the grouping is done in a way that formations of groups of **Directors** and **Star1** are formed.

In [15]:
# groupby on multiple cols
duo = movies.groupby(['Director', 'Star1'])
# size
duo.size()
# get_group
duo.get_group(('Aamir Khan', 'Amole Gupte'))

Series_Title Released_Year  Runtime  Genre  IMDB_Rating    Director  \
65  Taare Zameen Par          2007      165  Drama          8.4  Aamir Khan   

          Star1  No_of_Votes      Gross  Metascore  
65  Amole Gupte       168895  1223869.0        NaN

### **Task 12**: Find the most earning actor-director combo.

In [20]:
duo['Gross'].sum().sort_values(ascending=False).head(1)

Director        Star1         
Akira Kurosawa  Toshirô Mifune    2.999877e+09
Name: Gross, dtype: float64

### **Task 13**: FInd the best actor-genre combo in terms of average of metascore.

In [31]:
actGen_grp = movies.groupby(['Genre', 'Star1'])
actGen_grp['Metascore'].mean().reset_index().sort_values(by='Metascore', ascending=False).head(1)

Genre          Star1  Metascore
163  Adventure  Peter O'Toole      100.0

In [52]:
duo.agg(
    {
        'Runtime': ['min', 'max', 'mean']
    }
)
# temp[temp.reset_index().Runtime['min'] != temp.reset_index().Runtime['max']]

Runtime            
                                       min  max   mean
Director            Star1                             
Aamir Khan          Amole Gupte        165  165  165.0
Aaron Sorkin        Eddie Redmayne     129  129  129.0
Abdellatif Kechiche Léa Seydoux        180  180  180.0
Abhishek Chaubey    Shahid Kapoor      148  148  148.0
Abhishek Kapoor     Amit Sadh          130  130  130.0
...                                    ...  ...    ...
Zaza Urushadze      Lembit Ulfsak       87   87   87.0
Zoya Akhtar         Hrithik Roshan     155  155  155.0
                    Vijay Varma        154  154  154.0
Çagan Irmak         Çetin Tekindor     112  112  112.0
Ömer Faruk Sorak    Cem Yilmaz         127  127  127.0

[898 rows x 3 columns]